quinn-proto/src/connection/pacing.rs

```diff
@@ -0,0 +1,133 @@
+//! Pacing of packet transmissions.
+
+use std::time::{Duration, Instant};
+
+use tracing::warn;
+
+/// A simple token-bucket pacer. The bucket starts full and has an adjustable capacity. Once the
+/// bucket is empty, further transmission is blocked. The bucket refills at a rate slightly faster
+/// than one congestion window per RTT.
+pub struct Pacer {
+    capacity: u64,
+    tokens: u64,
+    prev: Instant,
+}
+
+impl Pacer {
+    /// Obtains a new [`Pacer`].
+    pub fn new(capacity: u64, now: Instant) -> Self {
+        Self {
+            capacity,
+            tokens: capacity,
+            prev: now,
+        }
+    }
+
+    /// Record that a packet has been transmitted.
+    pub fn on_transmit(&mut self, packet_length: u16) {
+        self.tokens = self.tokens.saturating_sub(packet_length.into())
+    }
+
+    /// Return how long we need to wait before sending a packet.
+    ///
+    /// If we can send a packet right away, this returns `None`. Otherwise, returns `Some(d)`,
+    /// where `d` is the time before this function should be called again.
+    ///
+    /// The 5/4 ratio used here comes from the suggestion that N = 1.25 in the draft IETF RFC for
+    /// QUIC.
+    pub fn delay(
+        &mut self,
+        smoothed_rtt: Duration,
+        mtu: u16,
+        window: u64,
+        now: Instant,
+    ) -> Option<Instant> {
+        const MAX_PACING_DELAY_BYTES: u64 = (1 << 16) - 1;
```

Comment:
Sorry, but I'm still unclear on why we have this bound, and why this particular value was chosen. Can you elaborate in a comment?

Reply:
The value is actually arbitrary.  What would you suggest @Ralith?  The bound is to ensure that pacing does not delay transmission indefinitely.

Reply:
Under what circumstance is that a risk?

Note that `mtu` is limited by the maximum theoretical size of a UDP packet (~2^16) and `self.capacity` is limited to at most the initial congestion window, which is [typically even smaller](https://quicwg.org/base-drafts/draft-ietf-quic-recovery.html#name-initial-and-minimum-congest). As written, I don't think this limit will ever take effect.

If you're concerned about the "wait until a full burst can be sent" strategy, maybe we should compromise that by allowing a lower threshold (say, at least N packets)? Or alternatively, roll back to waiting for a single packet in lieu of hard performance data to support the optimization.

Reply:
In that case, I will just delete this check.